File: src/flipper/evaluate/evaluate.c
/*
 * FlipDoctor — result evaluator implementation.
 *
 * Thresholds match db/thresholds.json. The guiding rule matches the Python
 * core: never emit a false PASS or FAIL. Use RESPONDS for "alive but not fully
 * validated" and INCONCLUSIVE for "insufficient evidence, here's the next step".
 *
 * Sole author: Krishita Sanjay Choksi.
 */
#include "evaluate.h"

/* thresholds (see db/thresholds.json) */
#define BAT_VMIN 3.30f
#define BAT_VNOM_MIN 3.60f
#define BAT_VMAX 4.25f
#define BAT_HEALTH_FAIL 60
#define BAT_HEALTH_WARN 85
#define SUBGHZ_RSSI_FLOOR -110.0f
#define SUBGHZ_RSSI_CEIL 10.0f
#define NFC_FIELD_MIN 1
#define IR_MATCH_MIN 0.75f
#define SD_MIN_FREE_RATIO 0.02f

static FdResult mk(
    FdSubsystem sub,
    FdStatus st,
    FdConfidence cf,
    const char* msg,
    const char* guide,
    bool needs_acc) {
    FdResult r = {
        .subsystem = sub,
        .status = st,
        .confidence = cf,
        .message = msg,
        .guidance = guide,
        .needs_accessory = needs_acc};
    return r;
}

static FdResult eval_battery(const FdReadings* r) {
    if(!r->responds || r->voltage_v <= 0.0f)
        return mk(
            FD_SUB_BATTERY, FD_STATUS_INCONCLUSIVE, FD_CONF_LOW,
            "Could not read battery voltage.",
            "Re-run; if unreadable the fuel-gauge may need attention.", false);
    if(r->voltage_v < BAT_VMIN || r->voltage_v > BAT_VMAX)
        return mk(
            FD_SUB_BATTERY, FD_STATUS_FAIL, FD_CONF_HIGH,
            "Battery voltage out of safe range.",
            "Charge fully and re-run; persistent out-of-range can indicate a "
            "failing cell.", false);
    if(r->health_pct < BAT_HEALTH_FAIL)
        return mk(
            FD_SUB_BATTERY, FD_STATUS_FAIL, FD_CONF_MEDIUM,
            "Battery health severely degraded.",
            "Expect very short runtime; service recommended.", false);
    if(r->health_pct < BAT_HEALTH_WARN || r->voltage_v < BAT_VNOM_MIN)
        return mk(
            FD_SUB_BATTERY, FD_STATUS_WARN, FD_CONF_HIGH,
            "Battery usable but not fresh.",
            "Normal aging; expect shorter runtime than a new unit.", false);
    return mk(FD_SUB_BATTERY, FD_STATUS_PASS, FD_CONF_HIGH, "Battery healthy.", "", false);
}

static FdResult eval_sd(const FdReadings* r) {
    if(!r->detected)
        return mk(
            FD_SUB_SD, FD_STATUS_FAIL, FD_CONF_HIGH, "No SD card detected.",
            "Reseat the card; if still undetected try another card.", false);
    if(!r->write_read_match)
        return mk(
            FD_SUB_SD, FD_STATUS_FAIL, FD_CONF_HIGH,
            "SD write/read mismatch.",
            "Back up now; reseat or try another card.", false);
    if(r->bad_sectors > 0)
        return mk(
            FD_SUB_SD, FD_STATUS_WARN, FD_CONF_MEDIUM, "SD readable but aging.",
            "Back up and consider replacing the card.", false);
    if(r->free_ratio < SD_MIN_FREE_RATIO)
        return mk(
            FD_SUB_SD, FD_STATUS_WARN, FD_CONF_HIGH, "SD nearly full.",
            "Free up space to avoid write failures.", false);
    return mk(FD_SUB_SD, FD_STATUS_PASS, FD_CONF_HIGH, "SD reads/writes cleanly.", "", false);
}

static FdResult eval_subghz(const FdReadings* r) {
    if(!r->responds)
        return mk(
            FD_SUB_SUBGHZ, FD_STATUS_FAIL, FD_CONF_HIGH,
            "Sub-GHz transceiver did not respond.",
            "Reboot and re-run; if it persists the radio may need service.", false);
    if(r->rssi_dbm < SUBGHZ_RSSI_FLOOR || r->rssi_dbm > SUBGHZ_RSSI_CEIL)
        return mk(
            FD_SUB_SUBGHZ, FD_STATUS_WARN, FD_CONF_MEDIUM,
            "Radio responds but ambient RSSI is out of range.",
            "Often a noisy RF environment; move away and re-run.", false);
    if(r->loopback_supported && r->loopback_ok)
        return mk(
            FD_SUB_SUBGHZ, FD_STATUS_PASS, FD_CONF_HIGH,
            "Radio responds and self-loopback succeeded.", "", false);
    return mk(
        FD_SUB_SUBGHZ, FD_STATUS_RESPONDS, FD_CONF_MEDIUM,
        "Transceiver responds.",
        "Alive, not fully validated; a full check needs a reference signal.", false);
}

static FdResult eval_nfc(const FdReadings* r) {
    if(!r->responds)
        return mk(
            FD_SUB_NFC, FD_STATUS_FAIL, FD_CONF_HIGH, "NFC front-end did not respond.",
            "Reboot and re-run; if it persists the NFC hardware may need service.", false);
    if(r->field_amplitude < NFC_FIELD_MIN)
        return mk(
            FD_SUB_NFC, FD_STATUS_FAIL, FD_CONF_MEDIUM,
            "NFC chip responds but drives no detectable field.",
            "Keep metal away and re-run; the coil may be damaged.", false);
    return mk(
        FD_SUB_NFC, FD_STATUS_RESPONDS, FD_CONF_MEDIUM,
        "NFC field detected - the coil is energizing.",
        "Proves the coil drives a field, not that every card will read. Use a "
        "reference card for a stronger check.", false);
}

static FdResult eval_ir_tx(const FdReadings* r) {
    if(r->emitted)
        return mk(FD_SUB_IR_TX, FD_STATUS_PASS, FD_CONF_HIGH, "IR transmitter emitted.", "", false);
    return mk(
        FD_SUB_IR_TX, FD_STATUS_FAIL, FD_CONF_HIGH, "IR transmitter did not fire.",
        "Re-run; if it persists the emitter may need service.", false);
}

static FdResult eval_ir_rx(const FdReadings* r) {
    if(!r->ir_tx_ok)
        return mk(
            FD_SUB_IR_RX, FD_STATUS_INCONCLUSIVE, FD_CONF_MEDIUM,
            "Cannot test IR receive while the transmitter is down.",
            "Fix the IR transmitter first, then re-run.", false);
    if(r->loopback_match_ratio >= IR_MATCH_MIN)
        return mk(FD_SUB_IR_RX, FD_STATUS_PASS, FD_CONF_HIGH, "IR receiver captured the pattern.", "", false);
    if(r->loopback_match_ratio <= 0.0f)
        return mk(
            FD_SUB_IR_RX, FD_STATUS_FAIL, FD_CONF_HIGH, "IR receiver captured nothing.",
            "Emitter works, so likely the receive photodiode. Retry in a dim room.", false);
    return mk(
        FD_SUB_IR_RX, FD_STATUS_INCONCLUSIVE, FD_CONF_LOW, "IR receiver captured a partial pattern.",
        "Often environmental; retry in a dim room. Definitive check needs a known target.", false);
}

static FdResult eval_gpio(const FdReadings* r) {
    if(r->jumper_present) {
        if(r->jumper_loopback_ok)
            return mk(FD_SUB_GPIO, FD_STATUS_PASS, FD_CONF_HIGH, "GPIO jumper loopback passed.", "", true);
        return mk(
            FD_SUB_GPIO, FD_STATUS_FAIL, FD_CONF_HIGH, "GPIO jumper loopback failed.",
            "Recheck wiring against the guide; if correct, a pin may be faulty.", true);
    }
    if(r->self_check_ok)
        return mk(
            FD_SUB_GPIO, FD_STATUS_RESPONDS, FD_CONF_LOW, "GPIO pins respond (weak self-check).",
            "Connect the guided loopback jumper for a definitive result.", true);
    return mk(
        FD_SUB_GPIO, FD_STATUS_INCONCLUSIVE, FD_CONF_LOW, "GPIO self-check inconclusive.",
        "Connect the guided loopback jumper and re-run.", true);
}

static FdResult eval_interactive(FdSubsystem sub, const FdReadings* r) {
    if(!r->interactive_done)
        return mk(
            sub, FD_STATUS_INCONCLUSIVE, FD_CONF_LOW, "Interactive check not completed.",
            "Run it on-device and confirm what you see/feel.", false);
    if(r->interactive_confirmed)
        return mk(sub, FD_STATUS_PASS, FD_CONF_HIGH, "All inputs/outputs confirmed.", "", false);
    return mk(
        sub, FD_STATUS_FAIL, FD_CONF_HIGH, "User reported a missing input/output.",
        "Note which specific button/pixel/LED/vibro failed.", false);
}

FdResult fd_evaluate(FdSubsystem sub, const FdReadings* r) {
    switch(sub) {
    case FD_SUB_BATTERY: return eval_battery(r);
    case FD_SUB_SD: return eval_sd(r);
    case FD_SUB_SUBGHZ: return eval_subghz(r);
    case FD_SUB_NFC: return eval_nfc(r);
    case FD_SUB_IR_TX: return eval_ir_tx(r);
    case FD_SUB_IR_RX: return eval_ir_rx(r);
    case FD_SUB_GPIO: return eval_gpio(r);
    case FD_SUB_BUTTONS:
    case FD_SUB_DISPLAY: return eval_interactive(sub, r);
    default:
        return mk(sub, FD_STATUS_INCONCLUSIVE, FD_CONF_LOW, "No evaluator.", "", false);
    }
}

File: flipdoctor/evaluate.py
"""The result evaluator: raw readings -> honest verdict + confidence.

This is the trust-critical component. Each subsystem has a small, explicit
rule set. The guiding principle: never emit a false PASS or FAIL. When a
self-test can only prove "alive", emit ``RESPONDS``; when the evidence is
insufficient, emit ``INCONCLUSIVE`` with a concrete next step.

Every function takes a plain ``readings`` dict (produced on-device by the HAL,
or by the fault simulator off-device) and returns a :class:`SubsystemResult`.
"""

from __future__ import annotations

from typing import Any, Dict, List

from .models import Confidence, Status, Subsystem, SubsystemResult
from .thresholds import load_thresholds, match_known_issue


def _with_known_issue(result: SubsystemResult) -> SubsystemResult:
    note = match_known_issue(result.subsystem.value, result.readings)
    if note:
        result.guidance = (result.guidance + " " + note).strip()
    return result


# --------------------------------------------------------------------------
# Per-subsystem evaluators
# --------------------------------------------------------------------------
def _eval_subghz(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    t = th["subghz"]
    if not r.get("responds"):
        return SubsystemResult(
            Subsystem.SUBGHZ, Status.FAIL, Confidence.HIGH,
            "Sub-GHz transceiver did not respond.",
            "The CC1101 radio is not answering on SPI. This usually indicates a "
            "radio front-end fault. Re-run after a reboot; if it persists the "
            "device likely needs service.",
            readings=r,
        )
    if r.get("chip_id") != t["expect_chip_id"]:
        return SubsystemResult(
            Subsystem.SUBGHZ, Status.INCONCLUSIVE, Confidence.MEDIUM,
            f"Radio responded but reported an unexpected chip id "
            f"({r.get('chip_id')!r}).",
            "Could be a firmware/HAL mismatch rather than hardware. Update "
            "firmware and re-run.",
            readings=r,
        )
    rssi = r.get("rssi_dbm")
    if rssi is not None and not (t["rssi_floor_dbm"] <= rssi <= t["rssi_ceiling_dbm"]):
        return SubsystemResult(
            Subsystem.SUBGHZ, Status.WARN, Confidence.MEDIUM,
            f"Radio responds but ambient RSSI ({rssi} dBm) is out of the "
            "expected range.",
            "Often just a noisy RF environment. Move away from strong "
            "transmitters and re-run before suspecting hardware.",
            readings=r,
        )
    if r.get("loopback_supported") and r.get("loopback_ok"):
        return SubsystemResult(
            Subsystem.SUBGHZ, Status.PASS, Confidence.HIGH,
            "Radio responds and TX->RX self-loopback succeeded.",
            "", readings=r,
        )
    return SubsystemResult(
        Subsystem.SUBGHZ, Status.RESPONDS, Confidence.MEDIUM,
        "Transceiver responds and reports the expected chip.",
        "Self-test confirms the radio is alive. A full TX/RX validation needs "
        "a known reference signal; treat this as 'alive', not 'perfect'.",
        readings=r,
    )


def _eval_nfc(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    t = th["nfc"]
    if not r.get("responds"):
        return SubsystemResult(
            Subsystem.NFC, Status.FAIL, Confidence.HIGH,
            "NFC front-end did not respond.",
            "The ST25R NFC chip is not answering. Re-run after a reboot; if it "
            "persists the NFC hardware likely needs service.",
            readings=r,
        )
    amp = r.get("field_amplitude", 0)
    if amp < t["field_min_amplitude"]:
        return SubsystemResult(
            Subsystem.NFC, Status.FAIL, Confidence.MEDIUM,
            "NFC chip responds but is not driving a detectable field.",
            "The coil may be damaged. Keep metal away from the antenna and "
            "re-run; if the field stays absent, the coil likely needs service.",
            readings=r,
        )
    return SubsystemResult(
        Subsystem.NFC, Status.RESPONDS, Confidence.MEDIUM,
        "NFC field detected — the coil is energizing.",
        "This proves the coil is driving a field. It does NOT prove every card "
        "type will read; some cards are simply unsupported. For a stronger "
        "check, present a known reference card.",
        readings=r,
    )


def _eval_ir_tx(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    if r.get("emitted"):
        return SubsystemResult(
            Subsystem.IR_TX, Status.PASS, Confidence.HIGH,
            "IR transmitter emitted the test pattern.",
            "", readings=r,
        )
    return SubsystemResult(
        Subsystem.IR_TX, Status.FAIL, Confidence.HIGH,
        "IR transmitter did not drive the emitter.",
        "The IR LED driver is not firing. Re-run; if it persists the emitter "
        "likely needs service.",
        readings=r,
    )


def _eval_ir_rx(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    t = th["ir"]
    ratio = r.get("loopback_match_ratio", 0.0)
    if not r.get("ir_tx_ok", True):
        return SubsystemResult(
            Subsystem.IR_RX, Status.INCONCLUSIVE, Confidence.MEDIUM,
            "Cannot test IR receive because the transmitter self-test failed.",
            "Fix/confirm the IR transmitter first, then re-run the receiver "
            "test.", readings=r,
        )
    if ratio >= t["loopback_min_match_ratio"]:
        return SubsystemResult(
            Subsystem.IR_RX, Status.PASS, Confidence.HIGH,
            "IR receiver captured the emitted pattern (self-loopback).",
            "", readings=r,
        )
    if ratio <= 0.0:
        return SubsystemResult(
            Subsystem.IR_RX, Status.FAIL, Confidence.HIGH,
            "IR receiver captured nothing during self-loopback.",
            "The emitter works, so the fault is likely the IR receive "
            "photodiode. Retry in a dim room away from bright/IR light; if it "
            "still captures nothing, the receiver may need service.",
            readings=r,
        )
    return SubsystemResult(
        Subsystem.IR_RX, Status.INCONCLUSIVE, Confidence.LOW,
        f"IR receiver captured a partial pattern (match {ratio:.0%}).",
        "Partial capture is often environmental (ambient IR, angle). Retry in "
        "a dim room; a definitive check uses a known IR target.",
        readings=r,
    )


def _eval_gpio(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    # GPIO is the honest 'needs accessory' case.
    if r.get("jumper_present"):
        if r.get("jumper_loopback_ok"):
            return SubsystemResult(
                Subsystem.GPIO, Status.PASS, Confidence.HIGH,
                "GPIO loopback jumper test passed.",
                "", needs_accessory=True, readings=r,
            )
        return SubsystemResult(
            Subsystem.GPIO, Status.FAIL, Confidence.HIGH,
            "GPIO loopback jumper test failed — a pin did not read back "
            "its driven value.",
            "Double-check the jumper wiring against the on-screen guide and "
            "re-run. If wiring is correct, that pin may be faulty.",
            needs_accessory=True, readings=r,
        )
    # No jumper: we can only do a weak self-check.
    if r.get("self_check_ok"):
        return SubsystemResult(
            Subsystem.GPIO, Status.RESPONDS, Confidence.LOW,
            "GPIO pins respond to drive/read self-check.",
            "This is a weak check. For a definitive result, connect the guided "
            "loopback jumper and re-run.",
            needs_accessory=True, readings=r,
        )
    return SubsystemResult(
        Subsystem.GPIO, Status.INCONCLUSIVE, Confidence.LOW,
        "GPIO self-check was inconclusive.",
        "Connect the guided loopback jumper for a definitive GPIO result.",
        needs_accessory=True, readings=r,
    )


def _eval_sd(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    t = th["sd"]
    if not r.get("detected"):
        return SubsystemResult(
            Subsystem.SD, Status.FAIL, Confidence.HIGH,
            "No SD card detected.",
            "Reseat the microSD card and re-run. If still undetected, try a "
            "different card to isolate card vs. slot.",
            readings=r,
        )
    if not r.get("write_read_match", False):
        return SubsystemResult(
            Subsystem.SD, Status.FAIL, Confidence.HIGH,
            "SD write/read test mismatched — data did not come back "
            "intact.",
            "Back up your data now. Reseat or try another card; if mismatches "
            "persist across cards the slot may be at fault.",
            readings=r,
        )
    if r.get("bad_sectors", 0) > t["max_bad_sectors"]:
        return SubsystemResult(
            Subsystem.SD, Status.WARN, Confidence.MEDIUM,
            f"SD readable but reported {r.get('bad_sectors')} bad sector(s).",
            "The card is aging. Back up and consider replacing it.",
            readings=r,
        )
    if r.get("free_ratio", 1.0) < t["min_free_ratio"]:
        return SubsystemResult(
            Subsystem.SD, Status.WARN, Confidence.HIGH,
            "SD card is nearly full.",
            "Free up space to avoid write failures.", readings=r,
        )
    return SubsystemResult(
        Subsystem.SD, Status.PASS, Confidence.HIGH,
        "SD card reads and writes cleanly.",
        "", readings=r,
    )


def _eval_battery(r: Dict[str, Any], th: Dict[str, Any]) -> SubsystemResult:
    t = th["battery"]
    v = r.get("voltage_v")
    if v is None:
        return SubsystemResult(
            Subsystem.BATTERY, Status.INCONCLUSIVE, Confidence.LOW,
            "Could not read battery voltage.",
            "Re-run; if unreadable the fuel-gauge may need attention.",
            readings=r,
        )
    if v < t["voltage_min_v"] or v > t["voltage_max_v"]:
        return SubsystemResult(
            Subsystem.BATTERY, Status.FAIL, Confidence.HIGH,
            f"Battery voltage {v:.2f} V is out of the safe range.",
            "Charge fully and re-run. A persistently out-of-range voltage can "
            "indicate a failing cell or fuel-gauge.",
            readings=r,
        )
    health = r.get("health_pct", 100)
    if health < t["health_fail_pct"]:
        return SubsystemResult(
            Subsystem.BATTERY, Status.FAIL, Confidence.MEDIUM,
            f"Battery health estimated at ~{health}%.",
            "The cell is significantly degraded; expect very short runtime. "
            "Battery service recommended.",
            readings=r,
        )
    if health < t["health_warn_pct"] or v < t["voltage_nominal_min_v"]:
        return SubsystemResult(
            Subsystem.BATTERY, Status.WARN, Confidence.HIGH,
            f"Battery usable but not fresh (health ~{health}%, {v:.2f} V).",
            "Still works; expect shorter runtime than a new unit. This is "
            "normal aging, not a fault.",
            readings=r,
        )
    return SubsystemResult(
        Subsystem.BATTERY, Status.PASS, Confidence.HIGH,
        f"Battery healthy (~{health}%, {v:.2f} V).",
        "", readings=r,
    )


def _eval_interactive(sub: Subsystem, r: Dict[str, Any]) -> SubsystemResult:
    confirmed = r.get("user_confirmed")
    if confirmed is None:
        return SubsystemResult(
            sub, Status.INCONCLUSIVE, Confidence.LOW,
            f"{sub.label} check not completed.",
            "This is an interactive test. Run it on-device and confirm what "
            "you see/feel.",
            readings=r,
        )
    if confirmed:
        return SubsystemResult(
            sub, Status.PASS, Confidence.HIGH,
            f"{sub.label}: all inputs/outputs confirmed by the user.",
            "", readings=r,
        )
    return SubsystemResult(
        sub, Status.FAIL, Confidence.HIGH,
        f"{sub.label}: the user reported a missing input/output.",
        "Note which specific button/pixel/LED/vibro failed; that narrows the "
        "fault.",
        readings=r,
    )


_DISPATCH = {
    Subsystem.SUBGHZ: _eval_subghz,
    Subsystem.NFC: _eval_nfc,
    Subsystem.IR_TX: _eval_ir_tx,
    Subsystem.IR_RX: _eval_ir_rx,
    Subsystem.GPIO: _eval_gpio,
    Subsystem.SD: _eval_sd,
    Subsystem.BATTERY: _eval_battery,
}


def evaluate_subsystem(subsystem: Subsystem, readings: Dict[str, Any]) -> SubsystemResult:
    """Evaluate one subsystem's raw readings into an honest result."""
    th = load_thresholds()
    if subsystem in _DISPATCH:
        return _with_known_issue(_DISPATCH[subsystem](readings, th))
    if subsystem in (Subsystem.BUTTONS, Subsystem.DISPLAY):
        return _with_known_issue(_eval_interactive(subsystem, readings))
    raise ValueError(f"No evaluator for {subsystem}")


def evaluate_all(readings_by_subsystem: Dict[Subsystem, Dict[str, Any]]) -> List[SubsystemResult]:
    """Evaluate every provided subsystem, in canonical subsystem order."""
    order = list(Subsystem)
    return [
        evaluate_subsystem(sub, readings_by_subsystem[sub])
        for sub in order
        if sub in readings_by_subsystem
    ]
